R/conditions.R

```diff
@@ -0,0 +1,815 @@
+#' handler stack and looks for a handler that can deal with the
+#' condition. If it finds a throwable handler, it throws it to the
+#' function that established it (\code{\link{with_handlers}()}). If R
+#' finds an inplace handler, it executes it locally. Only if no
+#' handler can deal with the critical condition will R jump to
+#' top-level (see \code{\link{rst_abort}()}).
+#'
+#' @param handler A handler function that takes a condition as
+#'   argument.
+#' @param muffle Whether to muffle the condition after executing an
+#'   inplace handler. The signalling function must have established a
+#'   muffling restart. Otherwise, an error will be issued.
+#' @seealso \code{\link{with_handlers}()} for examples,
+#'   \code{\link{restarting}()} for another kind of inplace handler.
+#' @export
+thrown <- function(handler) {
+  structure(handler, class = c("thrown", "handler"))
+}
+#' @rdname thrown
+#' @export
+inplace <- function(handler, muffle = FALSE) {
+  if (muffle) {
+    handler_ <- function(c) {
+      handler(c)
+      rst_muffle(c)
+    }
+  } else {
+    handler_ <- handler
+  }
+  structure(handler_, class = c("inplace", "handler"))
+}
+
+#' Create a restarting handler.
+#'
+#' This constructor automates the common task of creating an
+#' \code{\link{inplace}()} handler that invokes a restart.
+#'
+#' Jumping to a restart point from an inplace handler has two
+#' effects. First, the control flow jumps to wherever the restart was
+#' established, and the restart function is called (with \code{...},
+#' \code{.args} or \code{.fields} as arguments). Execution resumes
+#' from the \code{\link{with_restarts}()} call. Secondly, the transfer
+#' of the control flow out of the function that signalled the
+#' condition means that the handler has dealt with the condition. Thus
+#' the condition will not be passed on to other potential handlers
+#' established on the stack.
+#'
+#' @param .restart The name of a restart.
+#' @param .fields A character vector specifying the fields of the
+#'   condition that should be passed as arguments to the restart. If
+#'   named, the names (except empty names \code{""}) are used as
+#'   argument names for calling the restart function. Otherwise the
+#'   the fields themselves are used as argument names.
+#' @param ...,.args Additional arguments passed on the restart
+#'   function. These arguments are evaluated only once and
+#'   immediately, when creating the restarting handler.
+#'
+#' @export
+#' @seealso \code{\link{inplace}()} and \code{\link{thrown}()}.
+#' @examples
+#' # This is a restart that takes a data frame and names as arguments
+#' rst_bar <- function(df, nms) {
+#'   stats::setNames(df, nms)
+#' }
+#'
+#' # This restart is simpler and does not take arguments
+#' rst_baz <- function() "baz"
+#'
+#' # Signalling a condition parameterised with a data frame
+#' fn <- function() {
+#'   with_restarts(cnd_signal("foo", foo_field = mtcars),
```

Comment:
See the local variants in https://github.com/jimhester/withr

i.e. you could write:

```R
fn <- function() {
  local_restarts(
    rst_bar = rst_bar,
    rst_baz = rst_baz
  ) 
  cnd_signal("foo", foo_field = mtcars)
}
```

(I'm not sure that's the possible)